### PR TITLE
Abstract GPIO and dummy 

### PR DESCRIPTION
**Pull request by mifrandir**
Improves GPIO in the following ways:
1. Groups interfaces into lib/io/gpio.hpp to reduce file clutter.
2. Groups real implementation in lib/io/real_gpio.[ch]pp to avoid accidentally importing the non-abstract version. (Open for better naming suggestions.)
3. Adds a set of dummy GPIO classes that (a) show that/how the interface works, and (b) allows for some trivial tests.
4. Abstracts away the hardware to allow for software only tests; for example one could implement a connected GPIO pair (or even a network but that seems overkill) to allow for simulations where we can set GPIO pins at will and see how the pod reacts. 

This approach, if approved, could be taken for I2C, ADC, SPI, and CAN. The latter sounds paritcularly interesting if we are to implement a simulated CAN bus.

### CODE REVIEW

**Comment by kshxtij** [comment]
hardware_gpio.cpp for naming maybe?

**Comment by kshxtij** [comment]
For CAN we can use `vcan` and the virtual can interface exposed by socketcan instead of having to write a can simulator ourself.

We aren't using SPI for anything this year, so don't need that just yet.

Agreed for I2C and ADC, however with I2C we need to prioritize the multiplexer testing right now so that's on the backburner. 

I'll implement an interface for PWM while the driver code is being written.

File: test/utils/dummy_gpio.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        case core::DigitalSignal::kLow:
+          std::cout << "low";
+          break;
+        case core::DigitalSignal::kHigh:
+          std::cout << "high";
+      }
+      std::cout << std::endl;
+      return io::GpioWriteResult::kSuccess;
+    });
+  testRead(dummy_gpio, 4, "read 4\n");
```
Comment (SnickeyX) [inline test/utils/dummy_gpio.cpp:65]: can we have 'read from 4' instead

**Review by SnickeyX — CHANGES_REQUESTED** [review]
Minor changes only. This makes the purpose of the reader and writer much clearer.

File: test/utils/dummy_gpio.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  auto dummy_gpio_writer_opt = dummy_gpio.getWriter(pin);
+  ASSERT_TRUE(dummy_gpio_writer_opt);
+  auto dummy_gpio_writer = *dummy_gpio_writer_opt;
+  ASSERT_TRUE(dummy_gpio_writer);
+  const auto result = dummy_gpio_writer->write(state);
+  ASSERT_EQ(result, io::GpioWriteResult::kSuccess);
+  ASSERT_EQ(testing::internal::GetCapturedStdout(), expected_output);
+}
+
+TEST(DummyGpio, printToStderr)
```
Comment (SnickeyX) [inline test/utils/dummy_gpio.cpp:45]: since this is printToStderr, why arent we writing to cerr instead?

**Review by SnickeyX — CHANGES_REQUESTED** [review]

**Review by kshxtij — APPROVED** [review]
LGTM but i agree with renaming the test

File: lib/io/hardware_gpio.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Hardware GPIO interface, requires physical GPIO pins to be present. This should only
+ * be instantiated at the top level and then provided to users through the IGpio interface.
+ */
+class HardwareGpio {
+ public:
+  HardwareGpio(hyped::core::ILogger &log);
+
+  virtual std::optional<std::shared_ptr<IGpioReader>> getReader(const std::uint8_t pin);
+  virtual std::optional<std::shared_ptr<IGpioWriter>> getWriter(const std::uint8_t pin);
```
[the remark this answers is not in the document]
Comment (mifrandir) [reply lib/io/hardware_gpio.hpp:36]: Preference I suppose? Not sure we've agreed on one or the other.

[the remark this answers is not in the document]
Comment (SnickeyX) [reply lib/io/hardware_gpio.hpp:36]: Fair, just wanted to know if it was better practice to do this instead

**Review by SnickeyX — APPROVED** [review]

**Pull request merged**